[FLINK-21353][state] Add DFS-based StateChangelog (TM-owned state)

rkhachatryan opened the pull request:
## What is the purpose of the change

This is an alternative approach to implement FLINK-21353.
In contrast to #14839, it assumes TM-owned state which allows to simplify the code and avoid re-uploading state on abortion.

[Low-level design doc](https://docs.google.com/document/d/1vifa8cqZqirVr6Ke1A0FclLa2aNltOQb25GZon79btM/edit?usp=sharing)
[State ownership design doc](https://docs.google.com/document/d/1NJJQ30P27BmUvD7oa4FChvkYxMEgjRPTVdO1dHLl_9I/edit)

## Verifying this change

Added unit-tests:
- `BatchingStateChangeUploaderTest`
- `FsStateChangelogStorageLoaderTest`
- `FsStateChangelogStorageTest`
- `FsStateChangelogWriterSqnTest`
- `FsStateChangelogWriterTest`
- `RetryingExecutorTest`

Adjusted unit-tests to use DFS DSTL:
- `ChangelogDelegateEmbeddedRocksDBStateBackendTest`
- `ChangelogDelegateFileStateBackendTest`
- `ChangelogDelegateHashMapTest`
- `ChangelogDelegateMemoryStateBackendTest`

Existing integration and e2e tests should start using DFS DSTL after FLINK-21448.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

rkhachatryan commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:
```diff
@@ -0,0 +1,207 @@
+                results ->
+                        executor.execute(
+                                () -> {
+                                    results.forEach(e -> uploaded.put(e.sequenceNumber, e));
+                                    future.complete(buildHandle(results, readyToReturn));
```
Here, we need to check:
- whether a new checkpoint didn't already come in (need to resolve conflict which change to put into the `uploaded` set)
- whether the range wasn't truncated (so no need to safe, just complete the future)
- whether the checkpoint was not aborted

lqjack commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:
```diff
@@ -0,0 +1,207 @@
+
+    @Override
+    public void reset(SequenceNumber from, SequenceNumber to) {
+        LOG.debug("reset {} from {} to {}", logId, from, to);
```
What if we maintain the status of the changelogset ?  we can just change the status of them if we confirm or reset .

rkhachatryan replied:
I guess that's not relevant after adopting TM-side state ownership (`reset` is now no-op).

pnowojski commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/BatchingStateChangeUploader.java:
```diff
@@ -0,0 +1,175 @@
+    @GuardedBy("scheduled")
+    private ScheduledFuture<?> scheduledFuture;
+
+    private volatile Throwable error;
```
Why are you making this class a bit more difficult with `volatile` if it's already synchronised on the `scheduled` field? Can we move all of the synchronisation under the `scheduled` lock?

rkhachatryan replied:
Good point, I'll remove `volatile` and move accesses under `synchronized`.

pnowojski commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogCleanerImpl.java:
```diff
@@ -0,0 +1,66 @@
+        // callers if the cleanup doesn't keep up.
+        // In all cases except abort this are uploader threads; while on abort this is the task
+        // thread.
+        this(
+                new ThreadPoolExecutor(
+                        nThreads,
+                        nThreads,
+                        0L,
+                        TimeUnit.MILLISECONDS,
+                        new LinkedBlockingQueue<>(nThreads * MAX_TASKS_PER_THREAD),
+                        (ThreadFactory) Thread::new));
```
Let's not create yet another thread pool. `StreamTask#asyncOperationsThreadPool` sounds like a pool that is already doing IO against the same external system.

rkhachatryan replied:
Could you explain why not?

I see the following reasons to use a separate pool:
1. (as mentioned in the comment): use a blocking queue to provide back-pressure
2. It's easier to configure size with separate pools
3. Easier to debug

I'm assuming the overhead of having more threads per task bounded by `nThreads` negligible on modern OSes.

pnowojski replied:
I'm not worried about the overhead of having an idle thread pool.

1. I'm not sure if it's currently implemented, but having a backpressure on {{AsyncCheckpointRunnable}} also sounds reasonable.
2. If we have multiple pools doing very similar things, that should be tweaked in very similar way it's not easier to configure all of them independently for the user. At best user would need to know about each of those pools, and even assuming that he would have a problem if he wants to limits parallel threads writing to his DFS to for example 4, regardless if that's {{AsyncCheckpointRunnable}}, periodic materialisation or changelog writes.
3. Is it? What if you want to check if number of threads is an issue and you just simply want to bump/decrease the number of threads by a factor of 4? With single pool it's easier to do. With many pools you might hit a problem that you simply forgotten to modify that one thread pool that was actually causing the problem.

rkhachatryan replied:
I looked at the code again and I think that async cleanup is actually not necessary in this PR. It only happens on shutdown (and in that case it could be done by uploader threads).
In the previous PR it was called in more cases because of aborts.

So I'm going to remove it alltogether.

pnowojski commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogStorage.java:
```diff
@@ -0,0 +1,135 @@
+     * The log id is only needed on write to separate changes from different backends (i.e.
+     * operators) in the resulting file.
+     */
+    private transient AtomicInteger logIdGenerator = new AtomicInteger(0);
+
+    private transient volatile StateChangeUploader uploader;
+    private transient volatile FsStateChangelogCleaner cleaner;
+    private volatile ReadableConfig config; // todo: make final after FLINK-21804
```
Can you explain what is happening here? Why do we need 3 volatiles, `AtomicInteger` and `synchronized (this)`?

rkhachatryan replied:
All this was supposed to go away after FLINK-21804 (as I mentioned in the comment - but I should definitely have elaborate more).
FLINK-21804 is now merged, so I'll rework this part after rebasing the PR.

rkhachatryan replied:
After dropping 25d1aacca8af5b70f97ee3843c491ac0ffa7d6bd there is only `AtomicLong` here.
It is needed because this storage can be used by different Tasks.

pnowojski replied:
Thanks for the explanation. It makes sense to me now.

pnowojski commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:
```diff
@@ -0,0 +1,380 @@
+                        new UploadCompletionListener(keyGroupRange, range, readyToReturn, future));
+                if (!toUpload.isEmpty()) {
+                    uploader.upload(
+                            new UploadTask(
+                                    toUpload.values(),
+                                    this::handleUploadSuccess,
+                                    this::handleUploadFailure));
```
Do we need to make this class semi thread safe for handling the uploading results? Can not they go through mailbox for example? As far as I can tell only those two callbacks (`handleUploadSuccess` and `handleUploadFailure`) are called from different threads?

rkhachatryan replied:
I think it's doable and I left a comment about it:
```
    /** Lock to synchronize handling of upload completion with new upload requests. */
    // todo: replace with mailbox executor (after FLINK-23204)
    private final Object lock = new Object();
```
But as FLINK-23204 is currently not implemented (and there is no consensus on how to do that), I'd like not to add such a dependency.

pnowojski reviewed (commented):
Thanks @rkhachatryan for updating this PR. I've started to look at it and I've left a couple of preliminary comments. One larger one is if you can split commits in this PR into smaller easier to review chunks? For example almost every config option in the `FsStateChangelogOptions` sounds like could have been in its own commit. As it is, the PR is quite difficult to review due to its size and complexity.

rkhachatryan commented:
Thanks for reviewing @pnowojski ,

 I've rebased the PR and extracted 
- [hotfix][tests] Introduce TestTaskStateManagerBuilder 80c41901
- [FLINK-21353][state/changelog] Implement retries fe6aca62
- [FLINK-21353][state/changelog] Implement batching 0beffa13

I've also dropped 25d1aacca8af5b70f97ee3843c491ac0ffa7d6bd (which was a temporary initialization solution).

Please take a look.

pnowojski commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:
```diff
@@ -0,0 +1,380 @@
+    public SequenceNumber lastAppendedSequenceNumber() {
+        LOG.trace("query {} sqn: {}", logId, activeSequenceNumber);
+        SequenceNumber tmp = activeSequenceNumber;
+        rollover();
```
I remember we have been already discussing this `rollover()` in a method that sounds like a getter, but I forgot the rationale.
1. Can you remember me what's the reason behind it here?
2. At the very least maybe add a comment explaining it?
3. Maybe we can refactor the code to be more self-documenting without the need for the comment? Rename the method? Split it into two?

rkhachatryan replied:
So your question is why `rollover()` needs to be called here, right?

This is because the caller must be able to use the returned sequence number
to distinguish changes before and after this call. At the same time, we don't 
want to increment SQN on each append (to avoid too many objects and segments
in the resulting file).

I'll add a comment (unless you have a better idea of how to make it more clear).

pnowojski commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:
```diff
@@ -0,0 +1,380 @@
+ *
+ * <ol>
+ *   <li>Every change has at most one associated upload (retries are performed at a lower level)
+ *   <li>Every change is present in at most one collection: either uploaded OR notUploaded
```
nit `{@link #uploaded}` and `notUploaded`? (easier to keep the java doc up to date)

pnowojski replied:
what about `activeChangeSet`?

rkhachatryan replied:
What do you mean?
Whether there are there any invariants for `activeChangeSet`?
I wanted to emphasize that the change can not be present in both, as it's related to concurrency.

pnowojski commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:
```diff
@@ -0,0 +1,380 @@
+    @GuardedBy("lock")
+    private SequenceNumber lowestSequenceNumber = INITIAL_SQN;
+
+    /** Active changes, that will all use {@link #activeSequenceNumber}. */
+    private List<StateChange> activeChangeSet = new ArrayList<>();
+
+    /** {@link #activeChangeSet} size in bytes. */
+    private long activeChangeSetSize;
+
+    /** Changes that are not yet uploaded (upload not requested). */
+    private final NavigableMap<SequenceNumber, StateChangeSet> notUploaded = new TreeMap<>();
+
+    /** Uploaded changes, ready for use in snapshots. */
+    @GuardedBy("lock")
+    private final NavigableMap<SequenceNumber, UploadResult> uploaded = new TreeMap<>();
```
I'm lacking somewhere an explanation what is the relationship between activeChangeSet and notUploaded?

rkhachatryan replied:
The changes from `activeChangeSet` go to `notUploaded` in `rollover()` (I don't think this should be documented as such docs could easily become outdated, but I can add it if you prefer).

rkhachatryan replied:
I'll add a comment.
Renaming to something with "pending" I think would not improve clarity because there are multiple stages at which changes can be pending; and it's not clear what action is pending (sending to `Uploader` or moving to another collection).

pnowojski commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:
```diff
@@ -0,0 +1,380 @@
+        synchronized (lock) {
+            ensureCanPersist(from);
+            rollover();
+            Map<SequenceNumber, StateChangeSet> tailMap = notUploaded.tailMap(from, true);
+            Map<SequenceNumber, StateChangeSet> toUpload = new HashMap<>(tailMap);
+            tailMap.clear(); // prevent re-uploads
```
nit: `tailMap` variable is a bit confusing given that you are calling `tailMap` method twice here. Is the the intention to get the `tailMap(from, true)` from the `notUploaded` AND remove those elements from the `notUploaded` at the same time? And the issue is that there is no method like `NavigableMap#tailMapAndRemove`? If so, maybe a `private static` helper method here in this class to do just that?

rkhachatryan replied:
Good idea, I'll extract `drainTailMap` method.

pnowojski commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/StateChangeUploader.java:
```diff
@@ -0,0 +1,112 @@
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkState;
+
+// todo: consider using CheckpointStreamFactory / CheckpointStorageWorkerView
+//     Considerations:
+//     0. need for checkpointId in the current API to resolve the location
+//       option a: pass checkpointId (race condition?)
+//       option b: pass location (race condition?)
+//       option c: add FsCheckpointStorageAccess.createSharedStateStream
+//     1. different settings for materialized/changelog (e.g. timeouts)
+//     2. re-use closeAndGetHandle
+//     3. re-use in-memory handles (.metadata)
+//     4. handle in-memory handles duplication
```
Besides a `todo`, could you explain the purpose of this interface?

rkhachatryan replied:
The purpose of this interface is to abstract the different implementations of uploading state changelog parts.
It has a single `upload` method with a single `UploadTask` argument which is meant to inintiate such an upload.

rkhachatryan replied:
I'll add it though I think it contradicts to [Flink coding guidelines](https://flink.apache.org/contributing/code-style-and-quality-common.html#comments-and-code-readability) about redundancy.

pnowojski commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/StateChangeFsUploader.java:
```diff
@@ -0,0 +1,141 @@
+import static java.util.stream.Collectors.toList;
+import static org.apache.flink.core.fs.FileSystem.WriteMode.NO_OVERWRITE;
+
+class StateChangeFsUploader implements StateChangeUploader {
```
This is a synchronous implementation the `StateChangeUploader`, that is doing the IO operations in the `#upload()` call? If so, could you add this explanation in the java doc?

pnowojski reviewed (commented):
Thanks for the update and thanks a lot for splitting the PR. Now it's definitely much easier to read/review :)

A small batch of comments for a couple of first commits that I've reviewed so far (up to "[FLINK-21353][state] Add DFS-based StateChangelog").

pnowojski commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/RetryingExecutor.java:
```diff
@@ -0,0 +1,145 @@
+        }
+
+        private void handleError(Exception e) {
+            LOG.trace("execution attempt {} failed: {}", current, e.getMessage());
```
debug? info? But I would say almost for certain not just trace?

rkhachatryan replied:
Good piont, I'll change it to `info`.

pnowojski commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/RetryingExecutor.java:
```diff
@@ -0,0 +1,145 @@
+            if (!actionCompleted.get()) {
+                Optional<ScheduledFuture<?>> timeoutFuture = scheduleTimeout();
+                try {
+                    runnable.run();
```
How are you solving a (race?) condition when two attempts complete? Do you have a requirement that `RunnableWithException runnable` must be idempotent and you just ignore that issue?

rkhachatryan replied:
Yes, `runnable` must be idempotent, as multiple instances of it can be executed concurrently.
I'll add this requirement to the javadoc, thanks for pointing out.

rkhachatryan replied:
I've also found an issue in `UploadTask`. It was idempotent already (thanks to checks in writer); but not thread-safe (because `finished` wasn't thread safe). I've fixed it in 9bba8b004352b54e8d99277146708ed7525bff56.

pnowojski replied:
Can you introduce a separate interface for that? `IdempotentRunnable` or `IdempotentRunnableWithException`? I believe this is so important requirement that it should be marked better not just via a java doc.

Or maybe instead of using generic interface, define `RetriableAction` as a static interface in `RetryingExecutor`? I mean this would help to raise awareness of the idempotence requirement and also significantly improved IDE support (looking for usages, implementations etc).

pnowojski commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/RetryingExecutor.java:
```diff
@@ -0,0 +1,145 @@
+        private final ScheduledExecutorService executorService;
+        private final int current;
+        private final RetryPolicy retryPolicy;
+        private final AtomicBoolean actionCompleted;
+        private final AtomicBoolean attemptCompleted = new AtomicBoolean(false);
```
replace with a single lock? As it is it's not obvious what kind of race conditions are possible and how are you dealing with them.

rkhachatryan replied:
`actionCompleted` is shared across different attempts to execute an action.
Replacing it with `synchronized` would require a) adding a lock object per action; b) adding a shared boolean per action.
`attemptCompleted` would be changed from `final AtomicBoolean` to non-final `boolean`.

To me, this doesn't seem like an easier to understand solution.

However, I agree that the race conditions aren't obvious here and will describe them in the comments.

pnowojski reviewed (commented):
I've finished a full pass and I've left a couple of questions/comments.

Before the next pass, before creating new fixup commits, could you squash the already existing fixups (those that I have already seen)?

rkhachatryan commented on flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogOptions.java:
```diff
@@ -57,4 +87,18 @@
                                     + "(and not cleaned up by JM). "
                                     + "If the cleanup doesn't keep up then task might be back-pressured.");
+
+    public static final ConfigOption<Integer> NUM_UPLOAD_THREADS =
+            ConfigOptions.key("dstl.dfs.upload.num-threads")
+                    .intType()
+                    .defaultValue(5)
+                    .withDescription("Number of threads to use for upload.");
```
I think the assumption that async checkpoint part uploads are **always** using the same shared pool is incorrect. For example RocksDB uses separate threads to upload, and the IO executor is only used to wait for the result. 

The advantages are
- better isolation (e.g. heavy network traffic from the upstream does not affect snapshot uploading => checkpointing)
- fine-grained tuning is possible
- in some scenarios it's easier to debug (i.e. to see where changelog uploader is stuck, not why IO thread isn't uploading)

The disadvantages are flip-sides of the latter two: the additional config option and in more difficult to debug in other cases .

Additionally, here **scheduled** executor is needed, while current IO executor is just `Executor`.

rkhachatryan replied:
> This sounds like RocksDB issue 

I guess it was a design decision to (easily) throttle the number of request to DFS (S3). This is yet another reason for a separate thread pool here.

> I can see two reasons against using the same thread pool: ...2. having actions with different priority

It's not only about the priority, but also about isolation, as I wrote above. If there is an issue with uploading (e.g. bug or request throttling) then it will consume all IO threads - regardless of the priority. This makes it harder to debug because it will manifest itself in many different ways, giving false signals.

> I don't believe anyone will want to bother how much he should bump individual thread pools

I'd say users shouldn't bother at all and the defaults should be enough. But once there **is** some issue it might be helpful to see what are threads in each pool doing - and maybe adjust accordingly.

>  Besides I don't see this being an issue in this case either?

I looked at the code, and it seems can block releasing partitions. Maybe there are or will be some other effects?

> In that case it should be solved probably in a better way by still sharing thread pool, but with more sophisticated actions scheduler

I think it's better only because of the other point - ease of configuration. Otherwise, separate thread pools are simpler in code and more robust.

Besides that, async checkpointing phase (regardless of the backend) currently uses its own pool created in `StreamTask`, not the common IO pool.

rkhachatryan commented:
Thanks for the feedback, I've addressed or replied to your comments.
I've also found an issue in the `UploadTask` after your [comment](https://github.com/apache/flink/pull/15322#discussion_r669685980) (fixed in 9bba8b004352b54e8d99277146708ed7525bff56).

I've squashed existing fixups and rebased the PR, please take a look.

pnowojski reviewed (commented):
(feel free to squash the commits once more)

rkhachatryan commented:
@pnowojski,
do you have any further concerns?

pnowojski commented:
Yes. The thread pool issue for me is still not resolved. I'm currently busy with other issues, I will get back to this soon.

rkhachatryan commented:
@pnowojski, 
Do you still have any further concerns? :)

pnowojski reviewed (commented):
After some offline discussions. I can give a conditional LGTM, under the condition that we agree on getting rid of the technological debt with creating this new thread pool Soon™. For example after the feature freeze let's sit down and plan how the configuration of various thread pools should look like and this one would need to be the first one to adjust.

rkhachatryan commented:
If I remember those discussions correctly, we decided to approach thread pools refacrorings independently and as a whole. Not only this pool, and without blocking this PR.

The thread pool here should probably be merged with other upload pools, or with async pool, or with IO pool. This is to be decided.

I can agree that ideally there should be a common IO pool and we will try to merge the existing pools together as well as the pool here - but we should take all the factors into account.

pnowojski commented:
👍 Yes, that's what I had in mind

rkhachatryan commented:
Great, then I'll merge the PR once the CI completes.
(I've squashed the commits and rebased onto the latest master).

rkhachatryan commented:
Thanks for the review.
Merged into master as 13bfa2b670e8cbea81eccb36122491df0ead04f3..47446cd0a86343f55d327167fee199db62fdfbf4
